subprojects/dependency-management/src/main/java/org/gradle/api/internal/attributes/DefaultAttributesSchema.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.attributes;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.Maps;
+import org.gradle.api.Action;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.attributes.AttributeMatchingStrategy;
+import org.gradle.api.attributes.AttributesSchema;
+import org.gradle.api.attributes.HasAttributes;
+import org.gradle.internal.Cast;
+import org.gradle.internal.component.model.ComponentAttributeMatcher;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class DefaultAttributesSchema implements AttributesSchemaInternal {
+
+    private final Map<Attribute<?>, AttributeMatchingStrategy<?>> strategies = Maps.newHashMap();
+    private final Map<Key, List<? extends HasAttributes>> matchesCache = Maps.newHashMap();
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> AttributeMatchingStrategy<T> getMatchingStrategy(Attribute<T> attribute) {
+        AttributeMatchingStrategy<?> strategy = strategies.get(attribute);
+        if (strategy == null) {
+            throw new IllegalArgumentException("Unable to find matching strategy for " + attribute);
+        }
+        return Cast.uncheckedCast(strategy);
+    }
+
+    @Override
+    public <T> AttributeMatchingStrategy<T> attribute(Attribute<T> attribute) {
+        return attribute(attribute, null);
+    }
+
+    @Override
+    public <T> AttributeMatchingStrategy<T> attribute(Attribute<T> attribute, Action<? super AttributeMatchingStrategy<T>> configureAction) {
+        AttributeMatchingStrategy<T> strategy = Cast.uncheckedCast(strategies.get(attribute));
+        if (strategy == null) {
+            strategy = new DefaultAttributeMatchingStrategy<T>();
+            strategies.put(attribute, strategy);
+        }
+        if (configureAction != null) {
+            configureAction.execute(strategy);
+        }
+        return strategy;
+    }
+
+    @Override
+    public Set<Attribute<?>> getAttributes() {
+        return strategies.keySet();
+    }
+
+    @Override
+    public boolean hasAttribute(Attribute<?> key) {
+        return strategies.containsKey(key);
+    }
+
+    @Override
+    public List<? extends HasAttributes> getMatches(AttributesSchema producerAttributeSchema, List<HasAttributes> candidates, AttributeContainer consumer) {
```

Comment:
I had a discussion with @melix about this already. This method is only called during configuration time now. Should be fine without locking. However, this might change depending on how we proceed with variant selection in the future (see https://github.com/gradle/performance/issues/290)

What can be called now during execution are the matching methods of `DefaultArtifactTransformRegistrations` which are used by view creation through `incoming.getArtifacts(attributes)` (and related methods). That's why I did: https://github.com/gradle/gradle/pull/1032/commits/56a6560dbdc25b9a6da2975d183ce25ac39c87a7